### PR TITLE
PYTHON-4618 - Fix TypeError: Socket cannot be of type SSLSocket

### PR DESCRIPTION
**Pull request by NoahStapp**

### CODE REVIEW

File: pymongo/network_layer.py [hunk 1]
```diff
@@ -35,15 +36,31 @@
...
-    socket.sendall(buf)
+async def _async_sendall_ssl(sock: Union[socket.socket, _sslConn], buf: bytes) -> None:
+    view = memoryview(buf)
+    total_length = len(buf)
+    total_sent = 0
+    while total_sent < total_length:
+        try:
+            sent = sock.send(view[total_sent:])
+        except BLOCKING_IO_ERRORS:
+            await asyncio.sleep(0.5)
```
Comment (ShaneHarvey) [inline pymongo/network_layer.py:60]: Is there any way to poll here rather than sleep?

File: pymongo/network_layer.py [hunk 2]
```diff
@@ -35,15 +39,65 @@
...
+
+    while sent < len(buf):
+        try:
+            sent += sock.send(buf)
+        except BLOCKING_IO_ERRORS as exc:
+            fd = sock.fileno()
+            # Check for closed socket.
+            if fd == -1:
+                raise SSLError("Underlying socket has been closed") from None
+            if isinstance(exc, SSLWantReadError):
```
Comment (ShaneHarvey) [inline pymongo/network_layer.py:79]: Do these checks work for the pyopenssl exceptions?

File: pymongo/network_layer.py [hunk 3]
```diff
@@ -35,15 +39,65 @@
...
-def sendall(socket: Union[socket.socket, _sslConn], buf: bytes) -> None:
-    socket.sendall(buf)
+async def _async_sendall_ssl(
+    sock: Union[socket.socket, _sslConn], buf: bytes, loop: AbstractEventLoop
+) -> None:
+    fd = sock.fileno()
+    sent = 0
+
+    def _is_ready(fut: Future) -> None:
+        if fut.done():
```
Comment (ShaneHarvey) [inline pymongo/network_layer.py:65]: Should we remove reader/writer before checking done()?

File: pymongo/network_layer.py [hunk 4]
```diff
@@ -35,15 +39,65 @@
...
+        except BLOCKING_IO_ERRORS as exc:
+            fd = sock.fileno()
+            # Check for closed socket.
+            if fd == -1:
+                raise SSLError("Underlying socket has been closed") from None
+            if isinstance(exc, SSLWantReadError):
+                fut = loop.create_future()
+                loop.add_reader(fd, _is_ready, fut)
+                await fut
+            if isinstance(exc, SSLWantWriteError):
```
Comment (ShaneHarvey) [inline pymongo/network_layer.py:83]: How should we handle pyopenssl's WantX509LookupError?

File: pymongo/network_layer.py [hunk 5]
```diff
@@ -35,15 +39,65 @@
...
+                fut = loop.create_future()
+                loop.add_writer(fd, _is_ready, fut)
+                await fut
+
+
+async def _async_sendall_ssl_windows(sock: Union[socket.socket, _sslConn], buf: bytes) -> None:
```
Comment (ShaneHarvey) [inline pymongo/network_layer.py:89]: Could you add a comment that this is needed because the proactor event loop doesn't support add_writer/add_reader?

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: pymongo/network_layer.py (hunk 3 above)
Comment (NoahStapp) [reply pymongo/network_layer.py:65]: Good catch.

File: pymongo/network_layer.py (hunk 4 above)
Comment (NoahStapp) [reply pymongo/network_layer.py:83]: According to the docs, it should be called again with the same arguments: https://www.pyopenssl.org/en/23.0.0/api/ssl.html#OpenSSL.SSL.WantX509LookupError. Maybe we register both a reader and writer?

File: pymongo/network_layer.py (hunk 2 above)
Comment (NoahStapp) [reply pymongo/network_layer.py]: Fixed.

**Review by caseyclements — APPROVED** [review]
lgtm

File: pymongo/network_layer.py [hunk 6]
```diff
@@ -18,32 +18,105 @@
 import asyncio
 import socket
 import struct
+import sys
+from asyncio import AbstractEventLoop, Future
+from ssl import SSLError, SSLSocket
```
Comment (ShaneHarvey) [inline pymongo/network_layer.py:23]: This has to be wrapped with try/except to account for when the ssl module is not available.

File: pymongo/ssl_support.py [hunk 7]
```diff
@@ -53,6 +53,9 @@
     IPADDR_SAFE = True
     SSLError = _ssl.SSLError
     BLOCKING_IO_ERRORS = _ssl.BLOCKING_IO_ERRORS
+    BLOCKING_IO_READ_ERROR = _ssl.BLOCKING_IO_READ_ERROR
+    BLOCKING_IO_WRITE_ERROR = _ssl.BLOCKING_IO_READ_ERROR
```
Comment (ShaneHarvey) [inline pymongo/ssl_support.py:57]: BLOCKING_IO_READ_ERROR -> BLOCKING_IO_WRITE_ERROR

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

**Review by ShaneHarvey — APPROVED** [review]

**Pull request merged**